modules/core/src/main/java/org/locationtech/jts/index/strtree/STRtree.java

```diff
@@ -373,5 +402,91 @@ public Object nearestNeighbour(Envelope env, Object item, ItemDistance itemDist)
           ((ItemBoundable) minPair.getBoundable(1)).getItem()
       };
   }
+  private Object[] nearestNeighbour(BoundablePair initBndPair, double maxDistance, int k) 
+  {
+	double distanceLowerBound = maxDistance;
+    
+    // initialize internal structures
+    PriorityQueue priQ = new PriorityQueue();
+
+    // initialize queue
+    priQ.add(initBndPair);
+
+    java.util.PriorityQueue<BoundablePair> kNearestNeighbors = new java.util.PriorityQueue<BoundablePair>(k, new BoundablePairDistanceComparator(false));
+
+    while (! priQ.isEmpty() && distanceLowerBound >= 0.0) {
+      // pop head of queue and expand one side of pair
+      BoundablePair bndPair = (BoundablePair) priQ.poll();
+      double currentDistance = bndPair.getDistance();
+      
+      
+      /**
+       * If the distance for the first node in the queue
+       * is >= the current maximum distance in the k queue , all other nodes
+       * in the queue must also have a greater distance.
+       * So the current minDistance must be the true minimum,
+       * and we are done.
+       */
+      
+      
+      if (currentDistance >= distanceLowerBound){
+    	  break;  
+      }
+      /**
+       * If the pair members are leaves
+       * then their distance is the exact lower bound.
+       * Update the distanceLowerBound to reflect this
+       * (which must be smaller, due to the test 
+       * immediately prior to this). 
+       */
+      if (bndPair.isLeaves()) {
+        // assert: currentDistance < minimumDistanceFound
+    	
+    	  if(kNearestNeighbors.size()<k){
+	    	  	kNearestNeighbors.add(bndPair);
+    	  }
+    	  else
+    	  {
+
+    		  if(kNearestNeighbors.peek().getDistance()>currentDistance)
+    		  {
+    			  kNearestNeighbors.poll();
+    			  kNearestNeighbors.add(bndPair);
+    		  }
+    		  /*
+    		   * minDistance should be the farthest point in the K nearest neighbor queue.
+    		   */
+    		  distanceLowerBound = kNearestNeighbors.peek().getDistance();
+
+    	  }        
+      }
+      else {
+        // testing - does allowing a tolerance improve speed?
+        // Ans: by only about 10% - not enough to matter
+        /*
+        double maxDist = bndPair.getMaximumDistance();
+        if (maxDist * .99 < lastComputedDistance) 
+          return;
+        //*/
+
+        /**
+         * Otherwise, expand one side of the pair,
+         * (the choice of which side to expand is heuristically determined) 
+         * and insert the new expanded pairs into the queue
+         */
+        bndPair.expandToQueue(priQ, distanceLowerBound);
+      }
+    }
+    // done - return items with min distance
 
+    Object[] result = new Object[kNearestNeighbors.size()];
```

Comment:
@dr-jts No, it cannot. The K Nearest Neighbor Priority Queue needs to contain top K ItemBoundable (item+its bounding envelope). We have two options (1) Use the existing way (2) Maintain an additional list of top k nearest neighbors' bounding envelopes all the time.  The (1) is OK because normally the K is much smaller than the original dataset. The (2) is time-consuming because we need to update the additional list all the time along with KNN Priority Queue.